packages/subiquity_client/lib/src/server/launcher.dart

```diff
@@ -0,0 +1,233 @@
+import 'dart:io';
+import 'package:path/path.dart' as p;
+import 'package:xdg_directories/xdg_directories.dart' as xdg;
+
+import 'common.dart';
+import 'paths.dart';
+
+Future<int?> _readPidFile() async {
+  final file = _pidFile();
+  if (!await file.exists()) {
+    return null;
+  }
+  final content = await file.readAsString();
+  return int.tryParse(content.trim());
+}
+
+Future<void> _writePidFile(int pid) async {
+  final file = _pidFile();
+  try {
+    await file.create(recursive: true);
+    await file.writeAsString(pid.toString());
+  } on FileSystemException catch (e) {
+    log.warning('Error writing ${file.path} (${e.message}). '
+        'Hot restarts may cause multiple Subiquity test servers to run.');
+  }
+}
+
+File _pidFile() {
+  final dir = xdg.runtimeDir?.path ?? Directory.current.path;
+  return File('$dir/subiquity-test-server.pid');
+}
```

Comment:
we could drop these to a separate `pidfile.dart` while moving them :)

Reply:
I should have considered that before. It could be a small and self-contained PR. Anyways..